Added Jackson serializer/deserializer

bgorven opened the pull request:
This will allow VCards to be used with ObjectMapper, being transparently
marshalled to JCards and back.

mangstadt commented:
Buddy,

I don't understand what the "strict" checking is for in JCardRawReader.  Can you explain?

mangstadt commented:
Closing.  Does not provide enough of a benefit for the extra code.  Thank you for the suggestion, though.

bgorven commented:
Hi, sorry I didn't reply earlier. Turns out github notifications can be easy to miss.

The issue is that I'm supposed to be conforming to a json schema where vcards are included among json data as an array. The easiest way to do this would be to tell jackson how to serialize the ezvcard class, so that vcards included in pojos used by rest controllers would just be transparently serialized/deserialized. I tried to implement the serializer separately, but it doesnt work without changes to your code. The existing raw reader ignores tokens before the first left bracket, and doesn't consume the last bracket from the stream, which is great when a vcard's all you are looking for, but breaks down when jackson is expecting you to take exactly one array off the stream. Also both the reader and writer close the stream, which prevents the rest of the parent object from being processed.

bgorven commented:
The reason I put the strict flag in is because I didnt want to change the existing behavior, which people might be depending on, except when an existing parser is passed in (using the new constructor).

mangstadt commented on src/main/java/ezvcard/io/json/JCardRawReader.java:
```diff
@@ -86,11 +99,13 @@ public void readNext(JCardDataStreamListener listener) throws IOException {
 			if (prev == JsonToken.START_ARRAY && cur == JsonToken.VALUE_STRING && "vcard".equals(parser.getValueAsString())) {
 				break;
+			} else if (strict) {
+				throw new JCardParseException(JsonToken.VALUE_STRING, cur);
```
So, essentially, when "strict" is true, you are assuming that the next token is the beginning of a jCard.  Whereas, when strict is false, JCardRawReader will skip all input until it encounters a jCard or the end of the stream.

mangstadt replied:
This is not properly account for the case when the value is a string, but it's != "vcard"

bgorven replied:
That would be an invalid jcard, right? What kind of exception would you throw for that? Maybe add a String constructor to JCardParseException?

Another thing I've realized is that there might actually be a JsonToken.VALUE_NULL if the vcard field is null. I would expect that checking for that immediately after reading the first token and returning immediately (before the vcard is initialized in the listener with listener.beginVCard()) should be safe enough, like:
```java
JsonToken prev = parser.getCurrentToken();
if (prev == JsonToken.VALUE_NULL) {
	return;
}
```
Because if a string begins with `null`, there's nothing that could come after it and still be valid json.

mangstadt commented on src/main/java/ezvcard/VCard.java:
```diff
@@ -108,6 +113,8 @@
  * Represents a vCard.
  * @author Michael Angstadt
  */
+@JsonSerialize(using = JCardSerializer.class)
```
I am worried about the annotations here.  Will ez-vcard still work if a developer does not want to use the jCard functionality and excludes the jackson dependencies from the project?  Will there be a `ClassNotFoundException` when the developer tries to create a new `VCard`?

bgorven replied:
Removing them is no problem. The annotations could go on the getter/setter methods of the parent object, or the ObjectMapper can be configured to use the vcard serializers. My thinking was that you would need the dependency on jackson-databind regardless, as that's where the Serializer/Deserializer base classes are defined, but now you mention it, it could just as easily be an optional dependency. Anybody who needs this functionality would probably already have the dependency on jackson-databind in the first place.

mangstadt commented on src/main/java/ezvcard/io/json/JCardDeserializer.java:
```diff
@@ -0,0 +1,28 @@
+		} catch (JCardParseException e) {
+			throw new JsonParseException("Error parsing JCard: " + e.getMessage(), parser.getCurrentLocation());
+		} finally {
+			reader.close();
```
Is there a reason why the `JCardReader` object should be closed?  The way the code is now, calling `close()` has no effect because the `Reader` object inside of `JCardReader` is null.  I think this `close()` call should be removed.

bgorven replied:
Makes sense. That way there won't be any problem with closing the parser in the call to reader.close(). I guess I just always close any Closeable I create, out of habit.

mangstadt commented on src/main/java/ezvcard/io/json/JCardRawReader.java:
```diff
@@ -102,6 +117,10 @@ public void readNext(JCardDataStreamListener listener) throws IOException {
 		listener.beginVCard();
 		parseProperties();
+		
+		if (strict) {
```
I think this check should always be done, whether "strict" is true or not.

mangstadt commented on src/main/java/ezvcard/io/json/JCardRawReader.java:
```diff
@@ -271,6 +290,8 @@ public boolean eof() {
 	 * Closes the underlying {@link Reader} object.
 	 */
 	public void close() throws IOException {
-		reader.close();
```
Not sure why I'm not closing the parser here too.  Any reason why I shouldn't?

mangstadt commented:
No problem.  I think you can enable email notifications for this issue by clicking the "Subscribe" button on the right.

Please see the notes I've added to the code.  I've only looked at the deserialization part so far.  I'm still trying to wrap my head around this.  Thank you for your patience.

bgorven commented:
Thanks for your consideration. I've configured the correct email address in github now, so I shouldn't miss any more messages.

I've implemented the changes you suggested, added a couple of tests of the deserializer, and everything passed, except JCardRawReaderTest.ignore_other_json(), which tries to parse
```java
		"{" +
			"\"website\": \"example.com\"," +
			"\"vcard\":" +
				"[\"vcard\"," +
					"[" +
						"[\"prop\", {}, \"text\", \"value\"]" +
					"]," +
					"[" +
					"]" +
				"]" +
		"}"
```
and fails with the message `Expected END_ARRAY but was START_ARRAY`. It passes when I change it to 
```java
		"{" +
			"\"website\": \"example.com\"," +
			"\"vcard\":" +
				"[\"vcard\"," +
					"[" +
						"[\"prop\", {}, \"text\", \"value\"]" +
					"]" +
				"]" +
		"}";
```
so I've pushed it like that for now. Please take a look and let me know if it's ok.

mangstadt commented on src/main/java/ezvcard/io/json/JCardSerializer.java:
```diff
@@ -0,0 +1,68 @@
+	 *            the value that any new JCardSerializers will be configured
+	 *            with (defaults to true)O
+	 */
+	public static void addProdIdByDefault(boolean add) {
```
This method is not good because it sets the global state of the class.  Is it possible to configure the `ObjectMapper` in such a way so that it calls the `JCardSerializer(boolean)` constructor instead of the default constructor?  Or somehow get it to pass a boolean into the `serialize()` method?

bgorven replied:
I can add a boolean parameter to the configureObjectMapper static method, so if users choose to configure their ObjectMapper that way, it will use the chosen constructor, but if they have annotated their object with @JsonSerialize(using=JCardSerializer.class), I expect Jackson would be creating the deserializer with Class.newInstance(), so as I see it our options are:

- assign a default value in the no-arg constructor
- set the property to Boolean.valueOf(System.getProperty("ezvcard.json.addProdId"))
- use a global default value

If we go with the first option, users can always subclass JCardSerializer with something that calls the boolean constructor if they want different behavior than the default.

mangstadt replied:
The change you committed is exactly what I had in mind.  However, if we are allowing them to change the "addProdId" setting, we should allow them to change the other settings on the `JCardWriter` too.

* setIndent
* setScribeIndex
* setVersionStrict

I'm not sure how this can be done cleanly.  The `configureObjectMapper` method shouldn't have a million arguments.

Similarly, the `JCardDeserializer` should include the ability to set the following settings on the `JCardReader`:

* setScribeIndex

Thank you for your patience.  I want this to be perfect.

mangstadt commented:
Hi Buddy,

Were you notified of the notes I added to the code?  I made notes in these places:

* https://github.com/bgorven/ez-vcard/commit/2bedd65e7fcbb9e84dbef7a4105cb80143868562
* https://github.com/mangstadt/ez-vcard/pull/42/files

bgorven commented:
Hi, sorry, I did get the emails, sorry I haven't answered yet—I went down to a friend's wedding for the weekend, and but I haven't had a chance to sit down with the code yet. I should be able to answer most of your questions today.

mangstadt commented on src/main/java/ezvcard/io/json/JCardModule.java:
```diff
@@ -0,0 +1,135 @@
+
+		// Register the same scribe with both, as per the documentation of
+		// this.getScribeIndex()
+		ScribeIndex index = new ScribeIndex();
```
Because the serializer and deserializer are using the same `ScribeIndex` instance, this should be stored in a class field.  Then, the `registerScribe` method can add the scribe to this `ScribeIndex` instance (it doesn't have to call `registerScribe` on the serializer and deserializer).  The `getScribeIndex` would obviously return it.  And `setScribeIndex` would set it, along calling the `setScribeIndex` methods on the serializer and deserializer.  Does that make sense?

mangstadt commented on src/main/java/ezvcard/io/json/JCardFormat.java:
```diff
@@ -0,0 +1,48 @@
+public @interface JCardFormat {
+	/**
+	 * Whether or not to add a "PRODID" property to each vCard, saying that the
+	 * vCard was generated by this library. For 2.1 vCards, the extended
```
Remove the note about 2.1 vCards.  jCard only supports version 4.0.

mangstadt commented on src/main/java/ezvcard/io/json/JCardModule.java:
```diff
@@ -0,0 +1,135 @@
+
+	/**
+	 * Gets whether or not a "PRODID" property will be added to each vCard,
+	 * saying that the vCard was generated by this library. For 2.1 vCards, the
```
Remove the sentence about 2.1 vCards. jCard only supports version 4.0.

mangstadt commented on src/main/java/ezvcard/io/json/JCardModule.java:
```diff
@@ -0,0 +1,135 @@
+
+	/**
+	 * Sets whether or not to add a "PRODID" property to each vCard, saying that
+	 * the vCard was generated by this library. For 2.1 vCards, the extended
```
Remove the sentence about 2.1 vCards. jCard only supports version 4.0.

mangstadt commented on src/main/java/ezvcard/io/json/JCardSerializer.java:
```diff
@@ -0,0 +1,142 @@
+
+	/**
+	 * Gets whether or not a "PRODID" property will be added to each vCard,
+	 * saying that the vCard was generated by this library. For 2.1 vCards, the
```
Remove the sentence about 2.1 vCards. jCard only supports version 4.0.

mangstadt commented on src/main/java/ezvcard/io/json/JCardSerializer.java:
```diff
@@ -0,0 +1,142 @@
+
+	/**
+	 * Sets whether or not to add a "PRODID" property to each vCard, saying that
+	 * the vCard was generated by this library. For 2.1 vCards, the extended
```
Remove the sentence about 2.1 vCards. jCard only supports version 4.0.

mangstadt commented on src/test/java/ezvcard/io/json/JCardSerializerTest.java:
```diff
@@ -0,0 +1,159 @@
+
+		ObjectMapper mapper = getMapper();
+
+		final DefaultPrettyPrinter pp = new DefaultPrettyPrinter();
```
I didn't know about this pretty-printer class.  You probably noticed that in `JCardRawWriter` I tried to do my own pretty printing.  It doesn't work too well.  Do you think you could integrate jackson's pretty-printing system into `JCardRawWriter`?  Or is this something that can only be done at the "ObjectMapper" level?

bgorven replied:
I haven't had a minute to try it out, but as far as I can tell, you can set a pretty printer on the JsonGenerator, then as long as you don't call any writeRaw methods (currently the only place they are being called is in the indent method), it should all work as expected.

The only issue is that the pretty printer doesn't know what type of object it is serializing until it's done serializing it, so basically your only choices are to add a new line after every array element (as below), or not to indent jcards at all.

I mean, theoretically, it might be possible, when the json generator and pretty printer were both created by RawWriter, to switch out the array indenter depending on what was being serialized, but that may end up depending on jackson implementation details more than is wise. I'll try it, and if it works I'll submit it as a separate pull request, and you can decide on it then.

mangstadt replied:
That sounds like a good plan.  Thank you, Buddy.
